### PR TITLE
FCM client refactor

### PR DESCRIPTION
**Pull request by rowanseymour**
Moves to `Mailroom` instance alongside the elastic client, redis pool etc instead of being a global.

### CODE REVIEW

File: core/models/imports.go [hunk 1]
```diff
@@ -113,7 +113,7 @@ func (b *ContactImportBatch) tryImport(ctx context.Context, db *sqlx.DB, orgID O
 	}
 
 	// and apply in bulk
-	_, err = ApplyModifiers(ctx, db, nil, oa, modifiersByContact)
+	_, err = ApplyModifiers(ctx, db, nil, nil, oa, modifiersByContact)
```
Comment (rowanseymour) [inline core/models/imports.go:116]: imports don't send messages

File: web/surveyor/surveyor.go [hunk 2]
```diff
@@ -133,7 +133,7 @@ func handleSubmit(ctx context.Context, s *web.Server, r *http.Request) (interfac
 	if err != nil {
 		return nil, http.StatusInternalServerError, errors.Wrapf(err, "error starting transaction for session write")
 	}
-	sessions, err := models.WriteSessions(ctx, tx, s.RP, oa, []flows.Session{fs}, []flows.Sprint{sprint}, nil)
+	sessions, err := models.WriteSessions(ctx, tx, s.RP, nil, oa, []flows.Session{fs}, []flows.Sprint{sprint}, nil)
```
Comment (rowanseymour) [inline web/surveyor/surveyor.go:136]: surveyor sessions shouldn't be sending android messages

File: core/models/events.go [hunk 3]
```diff
@@ -237,7 +238,7 @@ func HandleAndCommitEvents(ctx context.Context, db QueryerWithTx, rp *redis.Pool
 }
 
 // ApplyModifiers modifies contacts by applying modifiers and handling the resultant events
-func ApplyModifiers(ctx context.Context, db QueryerWithTx, rp *redis.Pool, oa *OrgAssets, modifiersByContact map[*flows.Contact][]flows.Modifier) (map[*flows.Contact][]flows.Event, error) {
```
Comment (rowanseymour) [inline core/models/events.go:240]: modifiers don't send messages so don't need redis (for talking to courier) or the FCM client (for talking to Android)

**Review by nicpottier — APPROVED** [review]
I'm not sure this feels better to me, the hook interface changing seems especially ugly now, but if you really want it, go for it.

**Comment by rowanseymour** [comment]
I started to have my doubts realising how many places need to pass it around but I _think_ I still prefer the functional approach and if we're worried about function argument bloat we could possibly pack things like this into a context structure.

**Comment by nicpottier** [comment]
Does FCMClient really have any state that is interesting? Maybe our problem is having that be an object in the first place?

**Comment by nicpottier** [comment]
Looking at it it is basically just an http.Client, so it really isn't maintaining much state over just the config. I don't know, it isn't the end of the world for sure, but just really feels wrong to have the Hook stuff even know about FCMClient, especially given how rarely it is used.

**Comment by rowanseymour** [comment]
What's the difference between hooks knowing about redis.Pool and knowing about fcm.Client ?

**Comment by nicpottier** [comment]
To me it is that redis/sql are more generically used by other hooks. They also actually have state beyond just configs (pools for the connections).

**Comment by rowanseymour** [comment]
Ok so what do you want to do now that this is merged? I don't feel very strongly in favor of this so if you feel it's really wrong then I can revert it.

**Comment by nicpottier** [comment]
I approved it so I can live with it. :)

**Comment by rowanseymour** [comment]
I've reverted it for now so can ponder on it a bit more

**Pull request merged**